gusto/equations/shallow_water_equations.py

```diff
@@ -177,53 +155,38 @@ def __init__(self, domain, parameters, fexpr=None, bexpr=None,
 
         if bexpr is not None:
             topography = self.prescribed_fields('topography', domain.spaces('DG')).interpolate(bexpr)
-            if self.thermal:
-                n = FacetNormal(domain.mesh)
-                topography_form = subject(prognostic
-                                          (-topography*div(b*w)*dx
-                                           + jump(b*w, n)*avg(topography)*dS,
-                                           'u'), self.X)
-            else:
-                topography_form = subject(prognostic
-                                          (-g*div(w)*topography*dx, 'u'),
-                                          self.X)
+            topography_form = subject(prognostic
+                                      (-g*div(w)*topography*dx, 'u'),
+                                      self.X)
             residual += topography_form
 
-        # thermal source terms not involving topography.
-        # label these as the equivalent pressure gradient term
-        if self.thermal:
-            n = FacetNormal(domain.mesh)
-            source_form = pressure_gradient(subject(prognostic(-D*div(b*w)*dx
-                                                    - 0.5*b*div(D*w)*dx
-                                                    + jump(b*w, n)*avg(D)*dS
-                                                    + 0.5*jump(D*w, n)*avg(b)*dS,
-                                                    'u'), self.X))
-            residual += source_form
-
         # -------------------------------------------------------------------- #
         # Linearise equations
         # -------------------------------------------------------------------- #
         # Add linearisations to equations
         self.residual = self.generate_linear_terms(residual, self.linearisation_map)
 
+    def setup(self):
```

Comment:
Do we need this `setup` function? Could we just set `self.field_names` and `self.space_names` in the `__init__` method?

Reply:
same comment as above - happy to talk about alternatives because I don't like this either!

Reply:
Ah I understand, this is because the `ThermalShallowWaterEquations` inherit from the `ShallowWaterEquations`, and we go through the whole normal process of setting up the shallow water equations and then modify them...

It does feel like unnecessary effort to avoid calling the parent `__init__` method and duplicate all of the momentum equation in the `ThermalShallowWaterEquations` class.

If we required the list of field names and the dictionary of function spaces to be `@abstractproperty`s of all `PrognosticEquationSet`s would that work?